lib/Dialect/ESI/runtime/cpp/lib/backends/Cosim.cpp

```diff
@@ -0,0 +1,108 @@
+//===- Cosim.cpp - Connection to ESI simulation via capnp RPC -------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// DO NOT EDIT!
+// This file is distributed as part of an ESI package. The source for this file
+// should always be modified within CIRCT
+// (lib/dialect/ESI/runtime/cpp/lib/backends/Cosim.cpp).
+//
+//===----------------------------------------------------------------------===//
+
+#include "esi/backends/Cosim.h"
+
+#include "CosimDpi.capnp.h"
+#include <capnp/ez-rpc.h>
+
+#include <fstream>
+#include <iostream>
+
+using namespace esi;
+using namespace esi::backends::cosim;
+
+namespace esi::backends::cosim {
+
+/// Implement the SysInfo API for cosimulation.
+class CosimSysInfo final : public esi::SysInfo {
+private:
+  friend class CosimAccelerator;
+  CosimSysInfo() = default;
+
+public:
+  /// Get the ESI version number to check version compatibility.
+  uint32_t esiVersion() const override;
+
+  /// Return the JSON-formatted system manifest.
+  std::string rawJsonManifest() const override;
+};
+} // namespace esi::backends::cosim
+
+// For now, just return dummy values since these are not yet supported by the
+// hardware.
+uint32_t CosimSysInfo::esiVersion() const { return -1; }
+std::string CosimSysInfo::rawJsonManifest() const { return ""; }
+
+/// Parse the connection string and instantiate the accelerator. Support the
+/// traditional 'host:port' syntax and a path to 'cosim.cfg' which is output by
+/// the cosimulation when it starts (which is useful when it chooses its own
+/// port).
+std::unique_ptr<Accelerator>
+CosimAccelerator::connect(std::string connectionString) {
+  std::string portStr;
+  std::string host = "localhost";
+
+  size_t colon;
+  if ((colon = connectionString.find(':')) != std::string::npos) {
+    portStr = connectionString.substr(colon + 1);
+    host = connectionString.substr(0, colon);
+  } else {
+    std::ifstream cfg(connectionString);
+    std::string line, key, value;
```

Comment:
Would it make sense to make the connection string a JSON argument? Would be a bit more scaleable if other backends need different kinds of connection parameters, and definitely easier to read (instead of all this finicky string parsing).

Reply:
The problem with using JSON is that there's no JSON parser/library in the c++ standard library. Since (for now) we are distributing only the runtime source code (rather than distributing a compiled version) and expecting users to compile it themselves, I want to reduce the number of external dependencies. Even Python, pybind, and capnp are (theoretically) optional dependencies.

Reply:
In the longer term (when we get around to distributing compiled versions) for this API, I think a JSON connection string makes sense. Before then, I don't see connection strings being terribly complex -- host:port is probably the most complex thing we have to parse.